### PR TITLE
Project 4: Dependently Typed Language: Calculus of Constructions

### LINKED ISSUE #145: Project 4 Proposal: Dependently Typed Language
**What will you do?**
I would like to implement a dependently typed language. I will base the language off [the Calculus of Constructions](https://www.sciencedirect.com/science/article/pii/0890540188900053). If this ends up being too simple, I will also implement user defined inductive types by following [this paper](https://www.cs.cmu.edu/~fp/papers/mfps89.pdf) which shows how to represent them in CoC. Alternatively, I can use [the Calculus of Inductive Constructions](https://hal.inria.fr/hal-01094195/document) to directly implement this dependently typed language; I am unsure which approach would prove to be simpler.

**How will you do it?**
I will be following the papers linked above to implement the type system for this dependently typed language. I will start by creating a relatively simple parser for the language, then move on to implementing the inference rules for the type system given in the papers.

**How will you empirically measure success?**
Since Coq is based off the Calculus of Constructions, I should be able to write equivalent programs in Coq and the language I create and make sure they both type check. I would like to use this language for proofs; for example that `False -> True` for some encoding of `True`, `False`, and implication.

**Team members:**
Just me

### PR DESCRIPTION
**Pull request by chrisroman**
Closes #145 

Special thanks to Vincent Imbimbo (@Xaec6) for talking through and learning with me how CoC worked.

### CODE REVIEW

**Review by sampsyo — CHANGES_REQUESTED** [review]
Sounds interesting! While the end result didn't turn out to be all that implementation-y, there do seem to have been a few interesting insights here. It's not terribly surprising that substitution would be hard to get correct, for example. I would also have been interested to see a slightly deeper discussion of performance than just an anecdote that some type-checking operations seemed to be slow.

Anyway, I hope this was useful! You might want to consider linking to your implementation, which presumably is [this repository](https://github.com/chrisroman/coc).

File: content/blog/2019-12-18-dependently-typed-language/index.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+Two of the hardest things to get right were substitution and $\beta$-equivalence. These were the
+main reasons why I had bugs in my implementation. Additionally, debugging became extremely
+difficult as the programs increased in size simply because so many rules were being applied, it was
+hard to see exactly where things were going wrong. While CoC is a relatively simple language, some of
+the examples in Coquand and Huet's paper were very hard to follow, like the `inter` example. This made
+it more difficult to get an intuitive understanding of how CoC worked.
+
+## Further Work
+With more time, I would have liked to implement a general framework for creating inductive type
+in Coc explained by [this paper]((https://www.cs.cmu.edu/~fp/papers/mfps89.pdf)). With my implementation,
```
[the remark this answers is not in the document]
Comment (chrisroman) [reply content/blog/2019-12-18-dependently-typed-language/index.md]: Fixed!

File: content/blog/2019-12-18-dependently-typed-language/index.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+```
+
+`iff_refl` claims that if and only if is reflexive, while `iff_sym` claims that it is symmetric.
+`iff_sym` was actually a case where bugs in substitution came up, so I just used different names for
+abstractions/products.
+
+Interestingly, the last two theorems actually took about a second to typecheck, which is quite
+surprising because Coq can typecheck these terms extremely quickly. This may be because Coq is based on
+the Calculus of *Inductive* Constructions, which is quite different from CoC because it introduces
+universes. This increased expressivity probably allows for faster implementations.
```
[the remark this answers is not in the document]
Comment (chrisroman) [reply content/blog/2019-12-18-dependently-typed-language/index.md]: Yeah I wish I had some more time to do a rigorous profiling. Just now I tried a very simple profiling by writing code inside the function to get a total of how long it ran, but because there are lots of recursive calls it seemed like it was double counting a lot so I couldn't get a good estimate of where it was spending a lot of time...

**Comment by sampsyo** [comment]
Seems like you've addressed most of the stuff; thanks!! Remember to hit "request a review" when you're done.

**Review by sampsyo — APPROVED** [review]
Looks good!!
